### PR TITLE
[DO NOT MERGE][tt-train] Reduce Tile Problem Showcase

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Ticket
Link to Github Issue

### Problem description
While implementing cross_entropy_bw and softmax operations, I noticed consistent precision issues when summing across tiles using reduce_tile(). These issues can lead to incorrect outputs, especially when operating over wider tensors or with value ranges that amplify floating-point errors.

### What's changed
This PR introduces a targeted test suite designed to highlight the numerical inaccuracy of our current reduce_tile() implementation and to justify switching to a matmul-based reduction approach in precision-sensitive operations.

To investigate this further and provide clear evidence, I created a set of controlled experiments comparing:
1. reduce — Using reduce_tile() directly
Each tile is reduced via a single-threaded accumulation inside a register.
2. matmul — Using matmul_tiles() with a one-hot row vector
Input tiles are copied and accumulated using add_binary_tile(), and then reduced via matrix multiplication with a tile that has 1s in the first row and 0s elsewhere.

Test Setup
	•	Tensor shape: [1, 1, 1, W]
	•	Value range: randomized using uniform distributions ([0.0, 1.0] and [-1.0, 1.0])
	•	Number of tests: 1000
	•	Baseline: xt::xarray<float> used to compute reference results
	•	Output: JSONL file with raw results for plotting and further inspection
	
1. W = 32, uniform[0.0, 1.0]
One-tile comparison. Shows that even in simple cases, reduce_tile() introduces visible error.
Mse_plot((32(1 tile) unif(0 0, 1 0))

2. W = 1024, uniform[-1.0, 1.0]
Mse_plot_(1024(32 tiles) unif(-1 0, 1 0))

3. W = 1024, uniform[0.0, 1.0]
Realistic range similar to softmax input. Highlights that matmul stays closer to the baseline.
Mse_plot(1024(32 tiles) unif(0 0, 1 0))

4. Raw value plots (W = 1024)
Direct output comparison between reduce, matmul, and the baseline.
Result(1024(32 tiles) unif(0 0, 1 0))

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_reader_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        // calculate the address of the first tile in the row
+        // start_row is the number of rows already processed in other cores
+        uint32_t idx = (start_row + i) * Wt;  // (take already processed rows + current row)*Wt(number of tiles in row)
+
+        cb_reserve_back(cb_input, Wt);
+        uint32_t l1_write_addr = get_write_ptr(cb_input);
+        for (uint32_t j = 0; j < Wt; ++j) {
+            noc_async_read_tile(idx + j, input_address_generator, l1_write_addr);
+            l1_write_addr += tile_bytes;
+            noc_async_read_barrier();
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_reader_kernel.cpp]: move barrier outside of this cycle

File: tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    const uint32_t tile_bytes = get_tile_size(cb_output);
+    const DataFormat data_format = get_dataformat(cb_output);
+
+    const InterleavedAddrGenFast</* is dram */ true> output_addr_generator = {
+        .bank_base_address = output_addr, .page_size = tile_bytes, .data_format = data_format};
+
+    uint32_t end_row = start_row + num_rows_to_process;
+
+    cb_wait_front(cb_output, onetile);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp]: move this inside cycle

File: tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    uint32_t l1_read_addr = get_read_ptr(cb_output);
+    for (uint32_t r = start_row; r < end_row; r++) {
+        uint32_t idx = r * Wt;
+        noc_async_write_tile(idx, output_addr_generator, l1_read_addr);
+        l1_read_addr += tile_bytes;
+        noc_async_write_barrier();
+    }
+    cb_pop_front(cb_output, onetile);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp]: move this inside cycle

File: tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    const InterleavedAddrGenFast</* is dram */ true> output_addr_generator = {
+        .bank_base_address = output_addr, .page_size = tile_bytes, .data_format = data_format};
+
+    uint32_t end_row = start_row + num_rows_to_process;
+
+    cb_wait_front(cb_output, onetile);
+    uint32_t l1_read_addr = get_read_ptr(cb_output);
+    for (uint32_t r = start_row; r < end_row; r++) {
+        uint32_t idx = r * Wt;
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp]: idx should not be multiplied by Wt

Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/reduce_row_test/device/kernels/dataflow/reduce_row_test_op_writer_kernel.cpp]: Good catch Roman, thanks!

**Comment by blozano-tt** [comment]
@vmelnykovTT This PR has been identified as stale and no longer relevant. It will be closed in 14 days unless you add a comment protesting closure.

**Reason:** 11/14 files (79%) no longer exist
